src/tracer.js

```diff
@@ -0,0 +1,258 @@
+
+import assert from 'assert';
+import ConstSampler from './samplers/const_sampler.js';
+import * as constants from './constants.js';
+import * as opentracing from 'opentracing';
+import * as opentracing_tags from './tags.js';
+import myLocalIp from 'my-local-ip';
+import NoopReporter from './reporters/noop_reporter.js';
+import pjson from '../package.json';
+import Span from './span.js';
+import SpanContext from './span_context.js';
+import TextMapCodec from './propagators/text_map_codec.js';
+import * as thrift from './thrift.js';
+import Utils from './util.js';
+
+export default class Tracer {
+    _serviceName: string;
+    _reporter: Reporter;
+    _sampler: Sampler;
+    _host: Endpoint;
+    _tracerTags: any;
+    _injectors: any;
+    _extractors: any;
+
+    constructor(serviceName: string,
+            reporter: Reporter = new NoopReporter(),
+            sampler: Sampler = new ConstSampler(false),
+            port: number  = 0,
+            tracerTags: any = {}) {
+        tracerTags.jaegerClient = `Node-${pjson.version}`;
+        this._tracerTags = tracerTags
+        this._serviceName = serviceName;
+        this._reporter = reporter;
+        this._sampler = sampler;
+        this._host = Utils.createEndpoint(serviceName, myLocalIp(), port);
+        this._injectors = {};
+        this._extractors = {};
+
+        let textCodec = new TextMapCodec(false);
+        this.registerInjector(opentracing.FORMAT_TEXT_MAP, textCodec);
+        this.registerExtractor(opentracing.FORMAT_TEXT_MAP, textCodec);
+
+        let httpCodec = new TextMapCodec(true);
+        this.registerInjector(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+        this.registerExtractor(opentracing.FORMAT_HTTP_HEADERS, httpCodec);
+    }
+
+    _startInternalSpan(
+            spanContext: SpanContext,
+            operationName: string,
+            startTime: number,
+            tags: any = {},
+            rpcServer: boolean): Span {
+
+        let firstInProcess = rpcServer || (spanContext.parentId == null);
+        let span = new Span(
+            this,
+            operationName,
+            spanContext,
+            startTime,
+            firstInProcess
+        );
+
+        if (opentracing_tags.SAMPLING_PRIORITY in tags) {
+            span.setSamplingPriority(tags[opentracing_tags.SAMPLING_PRIORITY])
+        }
+
+        span.addTags(tags);
+        return span;
+    }
+
+    _report(span: Span): void {
+        span.addTags(this._tracerTags);
+        this._reporter.report(span);
+    }
+
+    registerInjector(format: string, injector: Injector): void {
+        this._injectors[format] = injector;
+    }
+
+    registerExtractor(format: string, extractor: Extractor): void {
+        this._extractors[format] = extractor;
+    }
+
+    /**
+    * The method for creating a root or child span.
+    *
+    * @param {object} fields - the fields to set on the newly created span.
+    * @param {string} fields.operationName - the name to use for the newly
```

Comment:
square brackets in jsdoc mean optional, but a span must always have an operationName, so its not in square brackets.